Update HA page: unresponsive endpoint detection and node failure fixes (#2878)
neondatabase/website: content/docs/introduction/high-availability.md

- **Object storage**

  Your data's primary, long-term storage is in **cloud object storage**, with **99.999999999%** durability, protecting against data loss regardless of Pageserver or Safekeeper status.

#### Recap of storage recovery times

| Component      | Failure impact                                | Recovery mechanism                  | Recovery time |
| -------------- | --------------------------------------------- | ----------------------------------- | ------------- |
| Safekeeper     | WAL writes continue to other Safekeepers      | Automatic redistribution across AZs | Seconds       |

> | Pageserver     | Read requests automaticallyroute to secondary | Automatic failover to secondary     | Seconds       |

Reviewer: ```suggestion
| Pageserver     | Read requests automatically route to secondary | Automatic failover to secondary     | Seconds       |
```

| Object storage | No impact - 99.999999999% durability          | Multi-AZ redundancy built-in        | Immediate     |

## Compute resiliency

> The compute layer does not yet support traditional high availability, but is built for resiliency and quick recovery from failures. A Neon compute is stateless, meaning failures do not affect your data. In the most common compute failures, _your connection string stays the same and new connections are automatically routed to the recovered instance_. However, as with any stateless service, your application should be configured to reconnect automatically. Downtime usually lasts seconds.

Reviewer: ```suggestion
The compute layer is built for resiliency and quick recovery from failures. A Neon compute is stateless, meaning failures do not affect your data. In the most common compute failures, _your connection string stays the same and new connections are automatically routed to a newly instantiated compute instance_. However, as with any stateless service, your application should be configured to reconnect automatically. Downtime usually lasts seconds.
```

![VM restarting after failure](/docs/introduction/vm_fails.png)

#### Degraded endpoints

> If a compute endpoint is in a degraded state (repeatedly crashing and restarting rather than failing outright), we will detect and reattach it automatically, typically within 5 minutes. During this time, your application may experience intermittent connectivity.

Reviewer: Can we add something like "this failure scenario is not common" and/or "Work is underway to reduce recovery times for this type of failure.

Author: Actually, 5 minutes is the reduced time. It used to be 30 minutes. This is the [fix](urhttps://github.com/neondatabase/cloud/commit/81ef2d4f97a13e4aa020c73c1d205489d9ffb26cl) that brought it down. 

I've changed the wording to unresponsive and also make it clear 5 minutes is a threshold.

Unresponsive endpoints
If a compute endpoint becomes unhealthy or unresponsive, we will automatically detect and reattach it to a new compute after 5 minutes. Your application may experience connectivity issues until the endpoint is restored.

> #### Node failures

Reviewer: same comment as above.

Node failures can affect multiple customers simultaneously when a Kubernetes node becomes unavailable. The control plane will reschedule compute instances to other healthy nodes, a process that typically takes a few minutes. While your data remains safe during this process, compute availability will be impacted until rescheduling is complete.

> 

Reviewer: Should we cover AZ and region outages? The Twitter thread that prompted this update was about AZ outages.

Author: For sure. This was added in a follow-up commit, but revising now to:

Availability Zones are physically separate data centers within a cloud region. When an AZ becomes unavailable, compute instances in that AZ will be automatically rescheduled to healthy AZs. Recovery time typically takes 1-10 minutes, depending on node availability in the destination AZs. Your connection string remains the same, and new connections will be routed to the recovered instance.

#### Recap of compute recovery times

| Failure type      | Impact                             | Recovery mechanism                      | Recovery time   |
| ----------------- | ---------------------------------- | --------------------------------------- | --------------- |
| Postgres crash    | Brief interruption                 | Automatic restart                       | Seconds         |
| VM failure        | Brief interruption                 | VM recreation and endpoint reattachment | Seconds         |
| Degraded endpoint | Possible intermittent connectivity | Automatic detection and reattachment    | Up to 5 minutes |
| Node failure      | Compute unavailable                | Rescheduling to healthy nodes           | ~2 minutes      |

### Impact on session data after a failure?

> While your application should handle reconnections automatically, session-specific data like temporary tables, prepared statements, and the Local File Cache ([LFC](/docs/reference/glossary#local-file-cache)), which stores frequently accessed data, will not persist across a failover. As a result, queries may initially run more slowly until the Postgres memory buffers and cache are rebuilt.

Reviewer: We use "failover" above. Should we use "recovery"? For the sake of argument, why can't we call this section "Compute failover" as some have suggested and explain that failover in Neon's serverless architecture is a little different from traditional failover.

Author: let's try that!
